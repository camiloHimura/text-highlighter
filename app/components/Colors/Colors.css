.colorsContainer{
    display: flex;
}

.colorsContainer > div{
    width: 60px;
    height: 40px;
    cursor: pointer;
    margin-left: 10px;
    border-radius: 2px;
    transition: all .2s;
    border: 1px solid black;
}

.colorsContainer > div.selected{
    width: 40px;
    height: 40px;
    border-radius: 50%;
}

.colorsContainer h2{
    margin-left: 20px;
}
