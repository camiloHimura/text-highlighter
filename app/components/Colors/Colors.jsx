import React from 'react';
import PropTypes from "prop-types";
import {Container, Title} from "./styles.js"

function Colors(props){
  const {options, optionSelected, onClick, title = ''} = props;

  return  <div data-test='container' className="--flex --alignCenter">
              {options.map((color, key) => {
                  const selected = optionSelected === color ? 'selected': '';
                  
                  return <Container 
                            data-test='color'
                            className={selected}
                            key={`${key}-${color}`} 
                            style={{"background": color}}
                            onClick={() => onClick(color)}
                          >
                          </Container>
              })}

              <Title>{title}</Title>
          </div>

}

Colors.propTypes = {
    options: PropTypes.arrayOf(PropTypes.string).isRequired,
    optionSelected: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
}

export default Colors ;
