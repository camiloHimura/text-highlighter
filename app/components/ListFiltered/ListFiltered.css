.listFiltered{
    height: 20vh;
    padding: 6px;
    overflow: auto;
    border: 1px solid black;
}

.listFiltered > div{
    width: fit-content;
    padding: 2px 10px;
    margin-bottom: 6px;
    position: relative;
    transition: all .2s;
    overflow: hidden;
}

.listFiltered > div:hover{
    padding-left: 34px;
}

.listFiltered .close{
    top: 50%;
    left: 0px;
    width: 20px;
    height: 20px;
    border: none;
    outline: none;
    color: white;
    margin-top: -10px;
    position: absolute;
    background: black;
    transition: all .2s;
    transform: translateX(-20px)
}

.listFiltered > div:hover .close{
    transform: translateX(6px)
}