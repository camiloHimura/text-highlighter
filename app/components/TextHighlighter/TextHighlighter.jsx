import React, {useState, useRef} from 'react';
import {connect} from 'react-redux';
import PropTypes from "prop-types";
import {Container} from "./styles.js"

import {addTextAction} from "../../state/actions";

const mapStateToProps = state => {
  return {
        selectHighLighter: state.colors.selectHighLighter,
      }
}

const mapDispachToProps = dispatch => {
  return {
    addText: (text, color) => dispatch(addTextAction({text, color})),
  }
}

export function TextHighlighter(props){
  const {optionSelected = "", selectHighLighter, addText} = props;
  const inputTextarea = useRef();

  function selectedText(event){
    const {target} = event;
    const textSlected = target.value.substring(target.selectionStart, target.selectionEnd);
    addText(textSlected, selectHighLighter);
  }

  return  <Container className={optionSelected}>
            <textarea autoFocus={true} ref={inputTextarea} onSelect={selectedText}>
            </textarea>
          </Container>
}

TextHighlighter.propTypes = {
  optionSelected: PropTypes.string,
  selectHighLighter: PropTypes.string,
  addText: PropTypes.func,
}

export default connect(mapStateToProps, mapDispachToProps)(TextHighlighter);
