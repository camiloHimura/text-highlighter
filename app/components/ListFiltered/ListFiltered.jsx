import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import PropTypes from "prop-types";
import "./ListFiltered.css"

import {removeTextAction} from "../../state/actions";

const mapStateToProps = state => {
    return {
            textList: state.textList,
            sortAscending: state.textList.sortAscending,
        }            
}

const mapDispachToProps = dispatch => {
    return {
        removeTextAction: (text, color) => dispatch(removeTextAction({text, color})),
    }
}

function ListFiltered(props) {
    const {textList, selectFilter, sortAscending, removeTextAction} = props;
    let finalList = textList[selectFilter] || [];

    useEffect(() => {
        finalList = finalList.sort(sortBy(sortAscending))
        
    }, [finalList.length, sortAscending]);

    function sortBy(isAsending){
        return (a, b) => {
            a = a.toLocaleLowerCase();
            b = b.toLocaleLowerCase();

            if (a > b) { return isAsending ? -1: 1 }
            if (b > a) { return isAsending ? 1 : -1 }
            return 0;
        }
    }

    function removeText(text, color){
        removeTextAction(text, color)
    }

    return  <div className="listFiltered">
                {finalList.map((info, key) => {
                            return <div key={`${key}-${info}${selectFilter}`}style={{"background": selectFilter}}>
                                        <span>{info}</span>
                                        <button className="close" onClick={() => removeText(info, selectFilter)}>x</button>
                                    </div>
                        })
                }
            </div>

}

ListFiltered.propTypes = {
    textList: PropTypes.shape({
        sortAscending: PropTypes.bool.isRequired
    }),
    selectFilter: PropTypes.string.isRequired, 
    sortAscending: PropTypes.bool.isRequired, 
    removeTextAction: PropTypes.func.isRequired
}

export default connect(mapStateToProps, mapDispachToProps)(ListFiltered);
