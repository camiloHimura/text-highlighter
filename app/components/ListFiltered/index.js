import ListFiltered from "./ListFiltered.jsx";
export default ListFiltered;