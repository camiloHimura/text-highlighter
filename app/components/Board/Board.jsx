import React, { Fragment } from 'react';
import PropTypes from "prop-types";

import Colors from "../Colors";
import TextHighlighter from "../TextHighlighter";
import ListFiltered from "../ListFiltered";
import Sort from "../Sort";


export function Board(props){
    const {selectFilter, selectFilterColorAction, selectHighLighter, selectHighLighterColorAction, colors} = props;

    return  <Fragment>
                <Colors title="HighLighter"
                    options={colors}
                    optionSelected={selectHighLighter}
                    onClick={selectHighLighterColorAction}
                />
                
                <div className="--clear-20"></div>
                <TextHighlighter
                    optionSelected={selectHighLighter}
                />

                <div className="--clear-20"></div>
                <div className="--flex --flexBetween">
                    <Colors title="Filter"
                        options={colors}
                        optionSelected={selectFilter}
                        onClick={selectFilterColorAction}
                        />
                    <Sort/>
                </div>

                <div className="--clear-20"></div>
                <ListFiltered
                    selectFilter={selectFilter}
                />
            </Fragment>
}

Board.propTypes = {
    colors: PropTypes.arrayOf(PropTypes.string).isRequired,
    selectFilter: PropTypes.string.isRequired,
    selectHighLighter: PropTypes.string.isRequired,
    selectHighLighterColorAction: PropTypes.func.isRequired,
    selectFilterColorAction: PropTypes.func.isRequired,
}

export default Board;
