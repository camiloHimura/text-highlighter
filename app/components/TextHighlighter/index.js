import TextHighlighter from "./TextHighlighter.jsx";
export default TextHighlighter;