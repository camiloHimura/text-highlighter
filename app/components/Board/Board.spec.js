import React from 'react';
import Board from './Board';
import Colors from '../Colors';
import ListFiltered from '../ListFiltered';
import TextHighlighter from '../TextHighlighter';
import { shallow } from 'enzyme'


describe('Board component', () => {
  let Component;
  let props;

  beforeEach(() => {
    props = {
      colors: ['blue', 'red', 'green'],
      selectFilter: 'selectFilter test',
      selectHighLighter: 'selectHighLighter test',
      selectFilterColorAction: jest.fn(),
      selectHighLighterColorAction: jest.fn(),
    }
    Component = shallow(<Board {...props}/>);
  })
  it('take snapshot', () => {
    expect(Component).toMatchSnapshot();
  });

  it('has Colors', () => {
    expect(Component.find(Colors)).toHaveLength(2);
  });

  it('check first Colors props', () => {
    const Color = Component.find(Colors).first()
    const {options, optionSelected} = Color.props();
    Color.prop('onClick')();

    expect(options).toEqual(props.colors);
    expect(optionSelected).toEqual(props.selectHighLighter);
    expect(props.selectHighLighterColorAction).toHaveBeenCalled();
  });

  it('check second Colors props', () => {
    const Color = Component.find(Colors).at(1)
    const {options, optionSelected} = Color.props();
    Color.prop('onClick')();

    expect(options).toEqual(props.colors);
    expect(optionSelected).toEqual(props.selectFilter);
    expect(props.selectFilterColorAction).toHaveBeenCalled();
  });

  it('has TextHighlighter', () => {
    expect(Component.find(TextHighlighter)).toHaveLength(1);
  });

  it('check TextHighlighter props', () => {
    expect(Component.find(TextHighlighter).props().optionSelected).toBe(props.selectHighLighter);
  });

  it('has ListFiltered', () => {
    expect(Component.find(ListFiltered)).toHaveLength(1);
  });

  it('check ListFiltered props', () => {
    expect(Component.find(ListFiltered).props().selectFilter).toBe(props.selectFilter);
  });

})