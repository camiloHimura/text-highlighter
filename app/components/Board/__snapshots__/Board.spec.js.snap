// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component take snapshot 1`] = `
<Fragment>
  <Colors
    onClick={[MockFunction]}
    optionSelected="selectHighLighter test"
    options={
      Array [
        "blue",
        "red",
        "green",
      ]
    }
    title="HighLighter"
  />
  <div
    className="--clear-20"
  />
  <Connect(TextHighlighter)
    optionSelected="selectHighLighter test"
  />
  <div
    className="--clear-20"
  />
  <div
    className="--flex --flexBetween"
  >
    <Colors
      onClick={[MockFunction]}
      optionSelected="selectFilter test"
      options={
        Array [
          "blue",
          "red",
          "green",
        ]
      }
      title="Filter"
    />
    <Connect(Sort) />
  </div>
  <div
    className="--clear-20"
  />
  <Connect(ListFiltered)
    selectFilter="selectFilter test"
  />
</Fragment>
`;
